hal/src/electron/modem/mdm_hal.cpp

```diff
@@ -851,14 +842,67 @@ bool MDMParser::init(DevStatus* status)
     if (waitFinalResp() != RESP_OK) {
         goto failure;
     }
-#endif
-    if (status)
+#endif // SOCKET_HEX_MODE
+    // enable power saving
+    if (_dev.lpm != LPM_DISABLED) {
+        // enable power saving (requires flow control, cts at least)
+        sendFormated("AT+UPSV=%d\r\n", _power_mode);
+        if (RESP_OK != waitFinalResp())
+            goto failure;
+        if (_power_mode != 0) {
+            _dev.lpm = LPM_ACTIVE;
+        }
+    }
+    // All _dev items collected without error, populate status.
+    if (status) {
         memcpy(status, &_dev, sizeof(DevStatus));
+    }
+    if (_dev.dev == DEV_SARA_R410) {
+        // Force Power Saving mode to be disabled for good measure
+        sendFormated("AT+CPSMS=0\r\n");
+        if (RESP_OK != waitFinalResp())
+            goto failure;
+        sendFormated("AT+CPSMS?\r\n");
+        if (RESP_OK != waitFinalResp())
+            goto failure;
+        // Force eDRX mode to be disabled
+        // 18/23 hardware doesn't seem to be disabled by default
+        sendFormated("AT+CEDRXS?\r\n");
+        int edrxActs[4] = { -1, -1, -1, -1 };
+        if (RESP_ERROR == waitFinalResp(_cbCEDRXS, edrxActs)) {
+            goto reset_failure;
+        }
+        int resetNeeded = false;
+        for (int i = 0; i < 4; i++) {
+            if (edrxActs[i] != -1) {
+                sendFormated("AT+CEDRXS=3,%d\r\n", edrxActs[i]);
+                waitFinalResp(); // Not checking for error since we will reset either way
+                resetNeeded = true;
+            }
+        }
+        if (resetNeeded) {
+            goto reset_failure;
+        }
+    } // if (_dev.dev == DEV_SARA_R410)
     UNLOCK();
     return true;
+
 failure:
     UNLOCK();
+    return false;
 
+reset_failure:
+    // Don't get stuck in a reset-retry loop
+    static int resetFailureAttempts = 0;
+    if (resetFailureAttempts++ < 3) {
```

Comment:
I'd move this condition to where the `resetNeeded` variable is checked. Otherwise there's still a chance for the system to stuck in a modem reinitialization loop, because `MDMParser::init()` returns `false` and causes `cellular_init()` to fail even if the number of resets has exceeded the limit

Reply:
It's ok if it gets stuck in an initialization loop by default exiting through `failure`, although it still won't do that (read on).  This protects against the specific case where we want to force a modem reset due to CEDRXS commands failing/succeeding, and preventing the cellular registration from continuing to run anyway.  Also if `MDMParser ::init()` returns false, it does not prevent registration from being run because the return value is not checked.

Reply:
> it does not prevent registration from being run because the return value is not checked

It's an oversight in the implementation of the system layer and not something we should rely on in the HAL code. If that's okay for `AT+CEDRXS` to fail and not cause a modem reset after a certain number of attempts then why not to express that explicitly in the code?

Reply:
Express it as a comment?  Basically preventing cellular registration just gets us back to retrying init faster, and the timeout of 3 attempts is arbitrary but allows us to register and connect even with an error, since that error is persisting and might just be a fluke or bug in the modem.  Allowing to continue on eventually helps the device recover in odd situations we can't predict.

Reply:
I was thinking about making `MDMParser::init()` return `true` in case of an eDRX-related error to make it apparent that we don't want that error to prevent the modem from registering in the network, but, yes, at the very least, a comment that we expect the system to ignore the result of `cellular_init()` and proceed with the connection attempt would be helpful

Reply:
Added the comments in a shadowy commented code block 🕶